scripts/actions/src/manage-scopes.ts

```diff
@@ -0,0 +1,150 @@
+import {
+  DefaultPostLoginApi,
+  DefaultPostLoginEvent,
+} from '../types/auth0-actions'
+import { ManagementClient, Permission } from 'auth0'
+
+const auth0Sdk = require('auth0')
+
+// basic OAuth 2.0 scopes that any client is allowed to request
+const DEFAULT_SCOPES = ['openid', 'profile', 'email', 'offline_access']
+// whitelist of scopes accessible to applications that aren't on the application whitelist
+// (currently empty, but we may expand it in future...)
+const SCOPE_WHITELIST: string[] = []
+
+/**
+ * 1. Filters the scopes available on a user's access token, based on the
+ * application they are using to log in.
+ * 2. Adds the requested scopes to the `idToken` so they are readable by applications
+ *
+ * 1. Filtering scopes:
+ *
+ * For applications on the whitelist, all scopes are allowed. Otherwise, the
+ * rule only allows a restricted subset. This makes it easy for third party
+ * applications to use our Auth0 system for authentication, without allowing
+ * them to authorize resources that shouldn't be accessible outside of
+ * first-party applications.
+ *
+ * Because we're manually editing the `accessToken.scopes` property, Auth0 will
+ * no longer check that the user actually has the permissions that they are
+ * requesting. Instead, we need to use the Management Client to get the
+ * permissions attached to the user, and only include the union of <allowed
+ * permissions> and <requested permissions> in the access token.
+ *
+ * 2. Adding scopes to id token
+ *
+ * Single Page Applications don't have an easy way to inspect which scopes are
+ * present on the Access Token. The Access Token should be treated as an opaque
+ * string (even though it's technically possible to decode it if it's a JWT).
+ * Instead, we should place any data that's supposed to be accessible to the
+ * application into the ID Token.
+ *
+ * We restrict this to the subset of applications that are in fact SPA's, as
+ * other applications may use cookies to store Auth0 sessions, and adding a lot
+ * of scopes to the ID Token (which will also by necessity be added to the
+ * Access Token) will significantly increase cookie size (which may cause failed
+ * HTTP requests – total HTTP headers should be smaller than 8kb).
+ */
+exports.onExecutePostLogin = async (
+  event: DefaultPostLoginEvent,
+  api: DefaultPostLoginApi
+) => {
+  // Skip if the analogous Rule was executed first. You can get these IDs from the Rules
+  // page in the Auth0 dashboard (https://manage.auth0.com/#/rules). Separate values are for dev, staging and prod
+  const rules = api.rules
+  if (
+    rules.wasExecuted('rul_iFouQc7sxIAebmG9') ||
```

Comment:
Do we know if this will still work once rules are fully deprecated (like, will it be a no-op?), or will we need to remove this check?

Reply:
I'm not sure, but I'll remove it to be on the safe side. I'll do that once the actions have been up and running for about a week